src/job_cache/daemon_cache.cpp

```diff
@@ -716,6 +717,24 @@ int DaemonCache::run() {
         wcl::log::info("Unrecognized event on %d: events = %d", event.data.fd, event.events)();
       }
     }
+
+    // Check for timeouts
+    std::unordered_set<int> clients_to_close;
+    for (auto &client : message_senders) {
```

Comment:
We should be removing these from `message_senders` and `message_parsers` no? If not we could close something more than once

Reply:
That's why I use an unordered_set and close them in a loop after both.

Reply:
right, but you never remove the old element from the variables: `message_senders` and `message_parsers`. If these loops ever run twice without resetting them then all the closed clients would get closed again.

It looks like this only happens right before exiting `DaemonCache::run` and we _shouldn't_ ever be calling that function twice but that's not actually enforced anywhere

Reply:
close_client handles that clean up automatically because there are many spots where we close a client. Also this is run every 5 seconds or after every IO event

Reply:
ah, right